### PR TITLE
Fix/timeout errors

### PR DESCRIPTION
**Pull request by Blu-J**
fixes: #997

image

### CODE REVIEW

File: appmgr/src/action/docker.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@ use crate::util::Version;
 use crate::volume::{VolumeId, Volumes};
 use crate::{Error, ResultExt, HOST_IP};
 
-pub const NET_TLD: &'static str = "embassy";
```
Comment (Blu-J) [inline appmgr/src/action/docker.rs:23]: Fix for clippy

File: appmgr/src/action/docker.rs [hunk 2]
```diff
@@ -230,24 +231,24 @@ impl DockerAction {
         }
     }
 
-    pub fn uncontainer_name<'a>(name: &'a str) -> Option<(PackageId<&'a str>, Option<&'a str>)> {
```
Comment (Blu-J) [inline appmgr/src/action/docker.rs:233]: Fix for clippy

Comment (Blu-J) [reply appmgr/src/action/docker.rs:233]: image

File: appmgr/src/action/docker.rs [hunk 3]
```diff
@@ -230,24 +231,24 @@ impl DockerAction {
...
+    pub fn uncontainer_name(name: &str) -> Option<(PackageId<&'_ str>, Option<&'_ str>)> {
         let (pre_tld, _) = name.split_once(".")?;
-        if pre_tld.contains("_") {
+        if pre_tld.contains('_') {
```
Comment (Blu-J) [inline appmgr/src/action/docker.rs:236]: Fix clippy

File: appmgr/src/action/docker.rs (hunk 2 above)
Comment (Blu-J) [reply appmgr/src/action/docker.rs:233]: https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

File: appmgr/src/action/docker.rs [hunk 4]
```diff
@@ -230,24 +231,24 @@ impl DockerAction {
...
-        &'a self,
+    async fn docker_args(
+        &self,
         ctx: &RpcContext,
         pkg_id: &PackageId,
         pkg_version: &Version,
         volumes: &Volumes,
         allow_inject: bool,
-    ) -> Vec<Cow<'a, OsStr>> {
+    ) -> Vec<Cow<'_, OsStr>> {
```
Comment (Blu-J) [inline appmgr/src/action/docker.rs:251]: image

File: appmgr/src/action/docker.rs [hunk 5]
```diff
@@ -230,24 +231,24 @@ impl DockerAction {
...
             let (pkg, name) = name.split_once("_")?;
             Some((Id::try_from(pkg).ok()?.into(), Some(name)))
         } else {
             Some((Id::try_from(pre_tld).ok()?.into(), None))
         }
     }
 
-    async fn docker_args<'a>(
-        &'a self,
+    async fn docker_args(
```
Comment (Blu-J) [inline appmgr/src/action/docker.rs:244]: image
https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

File: appmgr/src/action/docker.rs [hunk 6]
```diff
@@ -126,12 +109,29 @@ impl DockerAction {
...
-        Ok(if res.status.success() || res.status.code() == Some(143) {
-            Ok(if let Some(format) = &self.io_format {
+        let res = match res {
+            Race::Done(x) => x,
+            Race::TimedOut => {
+                if let Some(id) = id {
+                    signal::kill(Pid::from_raw(id as i32), signal::SIGKILL)
+                        .with_kind(crate::ErrorKind::Docker)?;
+                }
+                return Ok(Err((143, "Timed out. Retrying soon...$".to_owned())));
```
Comment (dr-bonez) [inline appmgr/src/action/docker.rs:130]: Drop the `$`

**Review by dr-bonez — CHANGES_REQUESTED** [review]

File: appmgr/src/action/docker.rs [hunk 7]
```diff
@@ -126,12 +109,29 @@ impl DockerAction {
...
         let res = tokio::select! {
-            res = handle.wait_with_output() => res.with_kind(crate::ErrorKind::Docker)?,
-            res = timeout_fut => res?,
+            res = handle.wait_with_output() => Race::Done(res.with_kind(crate::ErrorKind::Docker)?),
+            res = timeout_fut => {
+                res?;
+                Race::TimedOut
+            },
         };
-        Ok(if res.status.success() || res.status.code() == Some(143) {
```
[the remark this answers is not in the document]
Comment (Blu-J) [reply appmgr/src/action/docker.rs:133]: Can we even get here? Going to put it back but just wondering?

File: appmgr/src/action/docker.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Blu-J) [reply appmgr/src/action/docker.rs:23]: Neither did I, I think this is new-ish based on the rust version now, someone got convinced it was redundant. Strange to see something hit stable so fast.

File: appmgr/src/action/docker.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (dr-bonez) [reply appmgr/src/action/docker.rs:133]: This section of code is also invoked by the run_main section of the manager. This means that the main runtime of a service is run through this function. When you click "stop" on an application and it gracefully exits from the sigterm we send, we don't want to issue a crash notification.

**Review by dr-bonez — APPROVED** [review]

**Pull request merged**